DOCSP-49078: Extended JSON (#619)
mongodb/docs-csharp: source/data-formats/extended-json.txt

See the following table to see a description of each format:

.. list-table::
   :header-rows: 1
   :stub-columns: 1
   :widths: 10 40

   * - Name
     - Description

   * - **Extended**
     - | Also known as the *canonical* format, this JSON representation avoids loss of
         BSON type information.
       | This format prioritizes type preservation at the loss of human-readability and
         interoperability with older formats.

   * - **Relaxed Mode**
     - | JSON representation that describes BSON documents with some type information loss.
       | This format prioritizes human-readability and interoperability at the loss of
         certain type information.

   * - **Shell**
     - | JSON representation that matches the syntax used in the MongoDB shell.
       | This format prioritizes compatibility with the MongoDB shell, which often uses
         JavaScript functions to represent types.

   * - **Strict**
     - | *Deprecated.* This representation is the legacy format that fully conforms to
         the `JSON RFC <http://www.json.org/>`__ which allows any JSON parser to read the type information.

.. _extended_json_example_section:

> .. note::
> 
>    The driver parses the ``$uuid`` Extended JSON type from a string to a
>    ``BsonBinary`` object of binary subtype 4. For more information about ``$uuid`` field
>    parsing, see the
>    :spec:`special rules for parsing $uuid fields </extended-json.rst#special-rules-for-parsing-uuid-fields>`
>    section in the extended JSON specification.

Reviewer: ```suggestion
.. note::

   The {+driver-short+} parses the ``$uuid`` Extended JSON type from a string to a
   ``BsonBinary`` object of binary subtype 4. For more information about ``$uuid`` field
   parsing, see the
   :spec:`special rules for parsing $uuid fields </extended-json/extended-json.md#special-rules-for-parsing-uuid-fields>`
   section in the extended JSON specification.
```

To learn more about JSON, BSON, and Extended JSON, see
`our article about JSON and BSON <https://www.mongodb.com/resources/basics/json-and-bson>`__

> and :manual:`Extended JSON </reference/mongodb-extended-json/>` in the {+mdb-server+} manual.

Reviewer: I would move this up to the end of the overview section.

Extended JSON Examples
~~~~~~~~~~~~~~~~~~~~~~

The following examples show a document containing an ObjectId, date, and long
number field represented in each Extended JSON format. Click the tab that
corresponds to the format of the example you want to see:

.. tabs::

   .. tab:: Extended
      :tabid: extended-format

      .. code-block:: json

         {
           "_id": { "$oid": "573a1391f29313caabcd9637" },
           "createdAt": { "$date": { "$numberLong": "1601499609" }},
           "numViews": { "$numberLong": "36520312" }
         }

   .. tab:: Relaxed Mode
      :tabid: relaxed-mode-format

      .. code-block:: json

         {
           "_id": { "$oid": "573a1391f29313caabcd9637" },
           "createdAt": { "$date": "2020-09-30T18:22:51.648Z" },
           "numViews": 36520312
         }

   .. tab:: Shell
      :tabid: shell-format

      .. code-block:: json

         {
           "_id": ObjectId("573a1391f29313caabcd9637"),
           "createdAt": ISODate("2020-09-30T18:22:51.648Z"),
           "numViews": NumberLong("36520312")
         }

>    .. tab:: Strict
>       :tabid: strict-format
> 
>       .. code-block:: json
> 
>          {
>            "_id": { "$oid": "573a1391f29313caabcd9637" },
>            "createdAt": { "$date": 1601499609 },
>            "numViews": { "$numberLong": "36520312" }
>          }

Reviewer: This tab can be just be deleted unless the format deprecation is related to a non-driver dependency?
```suggestion
   .. tab:: Strict (deprecated)
      :tabid: strict-format

      This format is deprecated for applications except for <whatever conditions>.
      
      .. code-block:: json

         {
           "_id": { "$oid": "573a1391f29313caabcd9637" },
           "createdAt": { "$date": 1601499609 },
           "numViews": { "$numberLong": "36520312" }
         }
```

Read Extended JSON
------------------

You can read an Extended JSON documents into a {+language+} object by using the
``BsonSerializer.Deserialize<T>()`` method. The following example reads an
Extended JSON document into a ``BsonDocument`` object:

.. io-code-block::

   .. input:: /includes/fundamentals/code-examples/ExtendedJson.cs
      :language: csharp
      :start-after: start-read-ejson
      :end-before: end-read-ejson
      :dedent:

   .. output::
      :visible: false

      { "_id" : { "$oid" : "573a1391f29313caabcd9637" }, "createdAt" : { "$date" : "1970-01-19T12:51:39.609Z" }, "numViews" : 36520312 }

Write Extended JSON
-------------------

> You can write an Extended JSON string by calling the ``ToJson()`` method on a
> ``BsonDocument`` object or custom class. You must specify a ``JsonWriterSettings`` object
> with the ``OutputMode`` property set to the desired Extended JSON format as a parameter to
> the ``ToJson()`` method.

Reviewer: Nit:
```suggestion
You can write an Extended JSON string by calling the ``ToJson()`` method on a
``BsonDocument`` object or custom class. You must specify a ``JsonWriterSettings`` object
with the ``OutputMode`` property set to the desired Extended JSON format as a parameter.
```

Consider the following custom class:

.. literalinclude:: /includes/fundamentals/code-examples/ExtendedJson.cs
   :language: csharp
   :start-after: start-custom-class
   :end-before: end-custom-class

> The following example outputs an instance of ``MyDocument`` in
> Extended JSON format by specifying the ``CanonicalExtendedJson`` value to the ``OutputMode``
> property:

Reviewer: Nit:
```suggestion
The following example outputs an instance of ``MyDocument`` in
Extended JSON format by specifying the ``CanonicalExtendedJson`` value as an ``OutputMode``
property:
```
